File: examples/cpp/source/neural_networks/neural_net_dense_distr.cpp
/* file: neural_net_dense_distr.cpp */
/*******************************************************************************
* Copyright 2014-2018 Intel Corporation
* All Rights Reserved.
*
* If this  software was obtained  under the  Intel Simplified  Software License,
* the following terms apply:
*
* The source code,  information  and material  ("Material") contained  herein is
* owned by Intel Corporation or its  suppliers or licensors,  and  title to such
* Material remains with Intel  Corporation or its  suppliers or  licensors.  The
* Material  contains  proprietary  information  of  Intel or  its suppliers  and
* licensors.  The Material is protected by  worldwide copyright  laws and treaty
* provisions.  No part  of  the  Material   may  be  used,  copied,  reproduced,
* modified, published,  uploaded, posted, transmitted,  distributed or disclosed
* in any way without Intel's prior express written permission.  No license under
* any patent,  copyright or other  intellectual property rights  in the Material
* is granted to  or  conferred  upon  you,  either   expressly,  by implication,
* inducement,  estoppel  or  otherwise.  Any  license   under such  intellectual
* property rights must be express and approved by Intel in writing.
*
* Unless otherwise agreed by Intel in writing,  you may not remove or alter this
* notice or  any  other  notice   embedded  in  Materials  by  Intel  or Intel's
* suppliers or licensors in any way.
*
*
* If this  software  was obtained  under the  Apache License,  Version  2.0 (the
* "License"), the following terms apply:
*
* You may  not use this  file except  in compliance  with  the License.  You may
* obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
*
*
* Unless  required  by   applicable  law  or  agreed  to  in  writing,  software
* distributed under the License  is distributed  on an  "AS IS"  BASIS,  WITHOUT
* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*
* See the   License  for the   specific  language   governing   permissions  and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of neural network training and scoring in the distributed processing mode
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-NEURAL_NET_DENSE_DISTR"></a>
 * \example neural_net_dense_distr.cpp
 */

#include "daal.h"
#include "service.h"
#include "neural_net_dense_distr.h"

using namespace std;
using namespace daal;
using namespace daal::algorithms;
using namespace daal::algorithms::neural_networks;
using namespace daal::services;

/* Input data set parameters */
const string trainDatasetFileNames[4] =
{
    "../data/distributed/neural_network_train_dense_1.csv", "../data/distributed/neural_network_train_dense_2.csv",
    "../data/distributed/neural_network_train_dense_3.csv", "../data/distributed/neural_network_train_dense_4.csv"
};
const string trainGroundTruthFileNames[4] =
{
    "../data/distributed/neural_network_train_ground_truth_1.csv", "../data/distributed/neural_network_train_ground_truth_2.csv",
    "../data/distributed/neural_network_train_ground_truth_3.csv", "../data/distributed/neural_network_train_ground_truth_4.csv"
};
string testDatasetFile     = "../data/batch/neural_network_test.csv";
string testGroundTruthFile = "../data/batch/neural_network_test_ground_truth.csv";

const size_t nNodes = 4;
const size_t batchSize = 100;
const size_t batchSizeLocal = batchSize / nNodes;

TensorPtr trainingData[nNodes];
TensorPtr trainingGroundTruth[nNodes];
prediction::ModelPtr predictionModel;
prediction::ResultPtr predictionResult;
training::ModelPtr trainingModel;
training::TopologyPtr topology[nNodes];
training::TopologyPtr topologyMaster;

/* Algorithms to train neural network */
SharedPtr<training::Distributed<step2Master> > net;
SharedPtr<training::Distributed<step1Local> > netLocal[nNodes];

void initializeNetwork();
void trainModel();
void testModel();
void printResults();

int main(int argc, char *argv[])
{
    checkArguments(argc, argv, 10, &trainDatasetFileNames[0], &trainDatasetFileNames[1], &trainDatasetFileNames[2], &trainDatasetFileNames[3],
                  &trainGroundTruthFileNames[0], &trainGroundTruthFileNames[1], &trainGroundTruthFileNames[2], &trainGroundTruthFileNames[3],
                  &testDatasetFile, &testGroundTruthFile);

    initializeNetwork();
    trainModel();
    testModel();
    printResults();

    return 0;
}

void initializeNetwork()
{
    /* Create stochastic gradient descent (SGD) optimization solver algorithm */
    SharedPtr<optimization_solver::sgd::Batch<> > sgdAlgorithm(new optimization_solver::sgd::Batch<>());
    sgdAlgorithm->parameter.batchSize = batchSizeLocal;
    /* Read training data set from a .csv file and create tensors to store input data */
    for (size_t node = 0; node < nNodes; node++)
    {
        trainingData[node] = readTensorFromCSV(trainDatasetFileNames[node]);
        trainingGroundTruth[node] = readTensorFromCSV(trainGroundTruthFileNames[node], true);
    }
    Collection<size_t> dataDims = trainingData[0]->getDimensions();

    /* Configure the neural network */
    topologyMaster = configureNet();
    net = SharedPtr<training::Distributed<step2Master> >(new training::Distributed<step2Master>(sgdAlgorithm));

    /* Initialize the neural network on master node */
    services::Collection<size_t> oneBatchDimensions = dataDims;
    oneBatchDimensions[0] = batchSizeLocal;
    net->initialize(oneBatchDimensions, *topologyMaster);

    for(size_t node = 0; node < nNodes; node++)
    {
        /* Configure the neural network */
        topology[node] = configureNet();

        /* Pass a model from master node to the algorithms on local nodes */
        training::ModelPtr trainingModel(new training::Model());
        trainingModel->initialize<float>(oneBatchDimensions, *(topology[node]));
        netLocal[node] = SharedPtr<training::Distributed<step1Local> >(new training::Distributed<step1Local>());
        netLocal[node]->input.set(training::inputModel, trainingModel);
    }
}

void trainModel()
{
    /* Create stochastic gradient descent (SGD) optimization solver algorithm */
    SharedPtr<optimization_solver::sgd::Batch<> > sgdAlgorithm(new optimization_solver::sgd::Batch<>());

    /* Set learning rate for the optimization solver used in the neural network */
    float learningRate = 0.001f;
    sgdAlgorithm->parameter.learningRateSequence = NumericTablePtr(new HomogenNumericTable<>(1, 1, NumericTable::doAllocate, learningRate));
    sgdAlgorithm->parameter.batchSize = batchSizeLocal;

    /* Set the optimization solver for the neural network training */
    net->parameter.optimizationSolver = sgdAlgorithm;

    /* Run the neural network training */
    size_t nSamples = trainingData[0]->getDimensions().get(0);
    for (size_t i = 0; i < nSamples - batchSizeLocal + 1; i += batchSizeLocal)
    {
        /* Compute weights and biases for the batch of inputs on local nodes */
        for (size_t node = 0; node < nNodes; node++)
        {
            /* Pass a training data set and dependent values to the algorithm */
            netLocal[node]->input.set(training::data, getNextSubtensor(trainingData[node], i, batchSizeLocal) );
            netLocal[node]->input.set(training::groundTruth, getNextSubtensor(trainingGroundTruth[node], i, batchSizeLocal));

            /* Compute weights and biases on local node */
            netLocal[node]->compute();

            /* Pass computed weights and biases to the master algorithm */
            net->input.add(training::partialResults, node, netLocal[node]->getPartialResult());
        }

        /* Update weights and biases on master node */
        net->compute();
        training::ModelPtr wbModel = net->getPartialResult()->get(training::resultFromMaster)->get(training::model);
        checkPtr((void *)wbModel.get());
        NumericTablePtr wb = wbModel->getWeightsAndBiases();

        /* Update weights and biases on local nodes */
        for (size_t node = 0; node < nNodes; node++)
        {
            netLocal[node]->input.get(training::inputModel)->setWeightsAndBiases(wb);
        }
    }
    /* Finalize neural network training on the master node */
    net->finalizeCompute();

    /* Retrieve training and prediction models of the neural network */
    training::ModelPtr trModel = net->getResult()->get(training::model);
    checkPtr((void *)trModel.get());
    predictionModel = trModel->getPredictionModel<float>();
}

void testModel()
{
    /* Read testing data set from a .csv file and create a tensor to store input data */
    TensorPtr predictionData = readTensorFromCSV(testDatasetFile);

    /* Create an algorithm to compute the neural network predictions */
    prediction::Batch<> net;

    /* Set the batch size for the neural network prediction */
    net.parameter.batchSize = predictionData->getDimensionSize(0);

    /* Set input objects for the prediction neural network */
    net.input.set(prediction::model, predictionModel);
    net.input.set(prediction::data, predictionData);

    /* Run the neural network prediction */
    net.compute();

    /* Print results of the neural network prediction */
    predictionResult = net.getResult();
}

void printResults()
{
    /* Read testing ground truth from a .csv file and create a tensor to store the data */
    TensorPtr predictionGroundTruth = readTensorFromCSV(testGroundTruthFile);

    printTensors<int, float>(predictionGroundTruth, predictionResult->get(prediction::prediction),
                             "Ground truth", "Neural network predictions: each class probability",
                             "Neural network classification results (first 20 observations):", 20);
}

File: examples/java/com/intel/daal/examples/neural_networks/NeuralNetDenseDistr.java
/* file: NeuralNetDenseDistr.java */
/*******************************************************************************
* Copyright 2014-2018 Intel Corporation
* All Rights Reserved.
*
* If this  software was obtained  under the  Intel Simplified  Software License,
* the following terms apply:
*
* The source code,  information  and material  ("Material") contained  herein is
* owned by Intel Corporation or its  suppliers or licensors,  and  title to such
* Material remains with Intel  Corporation or its  suppliers or  licensors.  The
* Material  contains  proprietary  information  of  Intel or  its suppliers  and
* licensors.  The Material is protected by  worldwide copyright  laws and treaty
* provisions.  No part  of  the  Material   may  be  used,  copied,  reproduced,
* modified, published,  uploaded, posted, transmitted,  distributed or disclosed
* in any way without Intel's prior express written permission.  No license under
* any patent,  copyright or other  intellectual property rights  in the Material
* is granted to  or  conferred  upon  you,  either   expressly,  by implication,
* inducement,  estoppel  or  otherwise.  Any  license   under such  intellectual
* property rights must be express and approved by Intel in writing.
*
* Unless otherwise agreed by Intel in writing,  you may not remove or alter this
* notice or  any  other  notice   embedded  in  Materials  by  Intel  or Intel's
* suppliers or licensors in any way.
*
*
* If this  software  was obtained  under the  Apache License,  Version  2.0 (the
* "License"), the following terms apply:
*
* You may  not use this  file except  in compliance  with  the License.  You may
* obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
*
*
* Unless  required  by   applicable  law  or  agreed  to  in  writing,  software
* distributed under the License  is distributed  on an  "AS IS"  BASIS,  WITHOUT
* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*
* See the   License  for the   specific  language   governing   permissions  and
* limitations under the License.
*******************************************************************************/

/*
 //  Content:
 //     Java example of neural network in the distributed processing mode
 ////////////////////////////////////////////////////////////////////////////////
 */

package com.intel.daal.examples.neural_networks;

import com.intel.daal.algorithms.neural_networks.*;
import com.intel.daal.algorithms.neural_networks.prediction.*;
import com.intel.daal.algorithms.neural_networks.training.*;
import com.intel.daal.algorithms.neural_networks.layers.LayerDescriptor;
import com.intel.daal.algorithms.neural_networks.layers.NextLayers;
import com.intel.daal.algorithms.neural_networks.layers.ForwardLayer;
import com.intel.daal.algorithms.neural_networks.layers.BackwardLayer;
import com.intel.daal.algorithms.neural_networks.layers.ForwardInputId;
import com.intel.daal.algorithms.neural_networks.layers.ForwardResultId;
import com.intel.daal.algorithms.neural_networks.layers.BackwardInputId;
import com.intel.daal.algorithms.neural_networks.layers.BackwardResultId;
import com.intel.daal.data_management.data.Tensor;
import com.intel.daal.data_management.data.HomogenTensor;
import com.intel.daal.data_management.data.HomogenNumericTable;
import com.intel.daal.data_management.data.NumericTable;
import com.intel.daal.data_management.data_source.DataSource;
import com.intel.daal.examples.utils.Service;
import com.intel.daal.services.DaalContext;

/**
 * <a name="DAAL-EXAMPLE-JAVA-NEURALNETWORKDISTR">
 * @example NeuralNetDenseDistr.java
 */
class NeuralNetDenseDistr {

    private static final int nNodes = 4;
    private static final int batchSize = 100;
    private static final int batchSizeLocal = batchSize / nNodes;

    /* Input data set parameters */
    private static final String[] trainDatasetFiles     = { "../data/distributed/neural_network_train_dense_1.csv",
                                                            "../data/distributed/neural_network_train_dense_2.csv",
                                                            "../data/distributed/neural_network_train_dense_3.csv",
                                                            "../data/distributed/neural_network_train_dense_4.csv" };
    private static final String[] trainGroundTruthFiles = { "../data/distributed/neural_network_train_ground_truth_1.csv",
                                                            "../data/distributed/neural_network_train_ground_truth_2.csv",
                                                            "../data/distributed/neural_network_train_ground_truth_3.csv",
                                                            "../data/distributed/neural_network_train_ground_truth_4.csv" };
    private static final String testDatasetFile      = "../data/batch/neural_network_test.csv";
    private static final String testGroundTruthFile  = "../data/batch/neural_network_test_ground_truth.csv";

    private static DistributedStep2Master net;
    private static DistributedStep1Local[] netLocal = new DistributedStep1Local[nNodes];

    private static TrainingModel[] trainingModel = new TrainingModel[nNodes];

    private static Tensor[] trainingData = new Tensor[nNodes];
    private static Tensor[] trainingGroundTruth = new Tensor[nNodes];

    private static TrainingTopology topology;
    private static TrainingTopology[] topologyLocal = new TrainingTopology[nNodes];

    private static PredictionModel predictionModel;
    private static PredictionResult predictionResult;
    private static DaalContext context = new DaalContext();

    public static void main(String[] args) throws java.io.FileNotFoundException, java.io.IOException {

        initializeNetwork();
        trainModel();
        testModel();
        printResults();

        context.dispose();
    }

    private static void initializeNetwork() throws java.io.FileNotFoundException, java.io.IOException {
        com.intel.daal.algorithms.optimization_solver.sgd.Batch sgdAlgorithm =
            new com.intel.daal.algorithms.optimization_solver.sgd.Batch(context, Float.class, com.intel.daal.algorithms.optimization_solver.sgd.Method.defaultDense);
        sgdAlgorithm.parameter.setBatchSize(batchSizeLocal);

        /* Read training data set from a .csv file and create tensors to store input data */
        for (int node = 0; node < nNodes; node++) {
            trainingData[node] = Service.readTensorFromCSV(context, trainDatasetFiles[node]);
            trainingGroundTruth[node] = Service.readTensorFromCSV(context, trainGroundTruthFiles[node], true);
        }

        /* Configure the neural network */
        net = new DistributedStep2Master(context, sgdAlgorithm);
        topology = NeuralNetConfiguratorDistr.configureNet(context);
        net.parameter.setOptimizationSolver(sgdAlgorithm);

        long[] sampleSize = trainingData[0].getDimensions();
        sampleSize[0] = batchSizeLocal;
        /* Initialize the neural network on master node */
        net.initialize(sampleSize, topology);

        for (int node = 0; node < nNodes; node++) {
            /* Configure the neural network */
            topologyLocal[node] = NeuralNetConfiguratorDistr.configureNet(context);

            netLocal[node] = new DistributedStep1Local(context);

            trainingModel[node] = new TrainingModel(context);
            trainingModel[node].initialize(Float.class, sampleSize, topologyLocal[node]);

            /* Pass a model from master node to the algorithms on local nodes */
            netLocal[node].input.set(DistributedStep1LocalInputId.inputModel, trainingModel[node]);
        }
    }

    private static void trainModel() throws java.io.FileNotFoundException, java.io.IOException {
        /* Create stochastic gradient descent (SGD) optimization solver algorithm */
        com.intel.daal.algorithms.optimization_solver.sgd.Batch sgdAlgorithm =
            new com.intel.daal.algorithms.optimization_solver.sgd.Batch(context, Float.class, com.intel.daal.algorithms.optimization_solver.sgd.Method.defaultDense);

        /* Set learning rate for the optimization solver used in the neural network */
        double[] learningRateArray = new double[1];
        learningRateArray[0] = 0.001;
        sgdAlgorithm.parameter.setLearningRateSequence(new HomogenNumericTable(context, learningRateArray, 1, 1));
        sgdAlgorithm.parameter.setBatchSize(batchSizeLocal);

        /* Set the optimization solver for the neural network training */
        net.parameter.setOptimizationSolver(sgdAlgorithm);

        /* Run the neural network training */
        int nSamples = (int)trainingData[0].getDimensions()[0];
        for (int i = 0; i < nSamples - batchSizeLocal + 1; i += batchSizeLocal) {
            /* Compute weights and biases for the batch of inputs on local nodes */
            for (int node = 0; node < nNodes; node++) {
                /* Pass a training data set and dependent values to the algorithm */
                netLocal[node].input.set(TrainingInputId.data, Service.getNextSubtensor(context, trainingData[node], i, batchSizeLocal));
                netLocal[node].input.set(TrainingInputId.groundTruth, Service.getNextSubtensor(context, trainingGroundTruth[node], i, batchSizeLocal));

                /* Compute weights and biases on local node */
                PartialResult partialResult = netLocal[node].compute();

                /* Pass computed weights and biases to the master algorithm */
                net.input.add(DistributedStep2MasterInputId.partialResults, node, partialResult);
            }

            /* Update weights and biases on master node */
            DistributedPartialResult result = net.compute();
            NumericTable wb = result.get(DistributedPartialResultId.resultFromMaster).get(TrainingResultId.model).getWeightsAndBiases();

            /* Update weights and biases on local nodes */
            for (int node = 0; node < nNodes; node++) {
                netLocal[node].input.get(DistributedStep1LocalInputId.inputModel).setWeightsAndBiases(wb);
            }
        }

        /* Finalize neural network training on the master node */
        TrainingResult result = net.finalizeCompute();

        /* Retrieve training and prediction models of the neural network */
        TrainingModel trainingModel = result.get(TrainingResultId.model);
        predictionModel = trainingModel.getPredictionModel(Float.class);
    }

    private static void testModel() throws java.io.FileNotFoundException, java.io.IOException {

        /* Read testing data set from a .csv file and create a tensor to store input data */
        Tensor predictionData = Service.readTensorFromCSV(context, testDatasetFile);

        /* Create an algorithm to compute the neural network predictions */
        PredictionBatch net = new PredictionBatch(context);

        long[] predictionDimensions = predictionData.getDimensions();
        net.parameter.setBatchSize(predictionDimensions[0]);

        /* Set input objects for the prediction neural network */
        net.input.set(PredictionTensorInputId.data, predictionData);
        net.input.set(PredictionModelInputId.model, predictionModel);

        /* Run the neural network prediction */
        predictionResult = net.compute();
    }

    private static void printResults() throws java.io.FileNotFoundException, java.io.IOException {

        /* Read testing ground truth from a .csv file and create a tensor to store the data */
        Tensor predictionGroundTruth = Service.readTensorFromCSV(context, testGroundTruthFile);

        /* Print results of the neural network prediction */
        Service.printTensors("Ground truth", "Neural network predictions: each class probability",
                             "Neural network classification results (first 20 observations):",
                             predictionGroundTruth, predictionResult.get(PredictionResultId.prediction), 20);
    }
}
